### PR TITLE
th-167: Implement WS Location service

### PR DESCRIPTION
**Pull request by mV22110702**
+ Back-end part (geolocation runtime cache service, driverService.getGeolocation)
+ Front-end part (useGeolocation hook, sendDriverGeolocation helper, types)

### CODE REVIEW

File: backend/src/libs/packages/socket/socket.service.ts [hunk 1]
```diff
@@ -1,27 +1,47 @@
 import { type FastifyInstance } from 'fastify/types/instance';
 import { Server as SocketServer } from 'socket.io';
 
+import { type GeolocationCacheService } from '~/libs/packages/geolocation-cache/geolocation-cache.package';
```
Comment (DmytroVoronkov) [inline backend/src/libs/packages/socket/socket.service.ts]: missing file extention

File: backend/src/packages/business/business.service.ts [hunk 2]
```diff
@@ -136,11 +136,11 @@ class BusinessService implements IService {
     return await this.driverService.create({ payload, businessId });
   }
 
-  public updateDriver({
+  public async updateDriver({
     driverId,
     payload,
   }: DriverUpdatePayload): Promise<DriverCreateUpdateResponseDto> {
-    return this.driverService.update({
+    return await this.driverService.update({
```
Comment (h0wter) [inline backend/src/packages/business/business.service.ts]: What's this await for?

File: backend/src/libs/packages/geolocation-cache/geolocation-cache.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { GeolocationCacheService } from './geolocation-cache.package.js';
+
+const geolocationCacheService = GeolocationCacheService.getInstance();
+
+export { geolocationCacheService };
+export { GeolocationCacheService } from './geolocation-cache.package.js';
```
Comment (denred) [inline backend/src/libs/packages/geolocation-cache/geolocation-cache.ts:1]: It seems redundant to export both an instance and a class, exporting just the class should suffice.

File: backend/src/libs/packages/socket/socket.service.ts [hunk 4]
```diff
@@ -1,27 +1,47 @@
...
+    geolocationCacheService,
+  }: {
+    app: FastifyInstance;
+    geolocationCacheService: GeolocationCacheService;
+  }): void {
     this.io = new SocketServer(app.server, {
       cors: { origin: '*' },
     });
 
+    this.geolocationCacheService = geolocationCacheService;
```
Comment (denred) [inline backend/src/libs/packages/socket/socket.service.ts:30]: this.geolocationCacheService = GeolocationCacheService.getInstance();

File: backend/src/packages/drivers/driver.service.ts [hunk 5]
```diff
@@ -22,14 +26,45 @@ class DriverService implements IService {
...
+  }: {
+    driverRepository: DriverRepository;
+    userService: UserService;
+    groupService: GroupService;
+    geolocationCacheService: GeolocationCacheService;
+  }) {
     this.driverRepository = driverRepository;
     this.userService = userService;
     this.groupService = groupService;
+    this.geolocationCacheService = geolocationCacheService;
```
Comment (denred) [inline backend/src/packages/drivers/driver.service.ts:45]: this.geolocationCacheService = GeolocationCacheService.getInstance();

File: backend/src/packages/business/business.service.ts (hunk 2 above)
Comment (albsbz) [reply backend/src/packages/business/business.service.ts]: I suppose it's eslint rule to add await even async always returns promise

File: backend/src/packages/drivers/drivers.ts [hunk 6]
```diff
@@ -1,16 +1,18 @@
...
 import { DriverService } from './driver.service.js';
 
 const driverRepository = new DriverRepository(database, schema.drivers);
-const driverService = new DriverService(
+const driverService = new DriverService({
   driverRepository,
   userService,
   groupService,
-);
+  geolocationCacheService,
```
Comment (denred) [inline backend/src/packages/drivers/drivers.ts]: If you imported the class in DriverService, you shouldn't need this parameter.

File: backend/src/libs/packages/socket/socket.service.ts (hunk 4 above)
Comment (mV22110702) [reply backend/src/libs/packages/socket/socket.service.ts:30]: It's better to use dependency injection pattern and pass instance of the service as a parameter, I suppose.

**Review by h0wter — APPROVED** [review]

**Review by DmytroVoronkov — APPROVED** [review]

**Review by denred — APPROVED** [review]

**Review by h0wter — APPROVED** [review]

**Review by denred — APPROVED** [review]

**Review by kalinkaaaa14 — APPROVED** [review]

**Review by PavloSukhinin — APPROVED** [review]

**Pull request merged**